dgl_sparse/include/sparse/sparse_format.h

```diff
@@ -18,42 +18,71 @@
 namespace dgl {
 namespace sparse {
 
-/** @brief SparseFormat enumeration */
+/** @brief SparseFormat enumeration. */
 enum SparseFormat { kCOO, kCSR, kCSC };
 
-/** @brief CSR sparse structure */
+/** @brief COO sparse structure. */
+struct COO {
+  /** @brief the dense shape of the matrix. */
+  int64_t num_rows = 0, num_cols = 0;
+  /** @brief COO format row array of the matrix. */
+  torch::Tensor row;
+  /** @brief COO format column array of the matrix. */
+  torch::Tensor col;
+  /** @brief data index tensor. When it is null, assume it is from 0 to NNZ - 1.
+   */
+  torch::optional<torch::Tensor> data;
```

Comment:
COO matrix shall not have `data` because the values are stored in the same order of row/col indices. To ensure there is no data corruption, you can check the `data` array of the old `COOMatrix` of DGL must be null after format conversion.

Reply:
I added `data` in COO to make it consistent with the `COOMatrix` of DGL. Is it possible that we create a sparse format from a CSC and somehow it goes though `CSC->CSR` and `CSR->COO` conversions. Does DGL guarantee the order of the final COO follow the same value order of the CSC?

Reply:
I got you point. In the case of `CSC->CSR` and `CSR->COO`, the final COO should be ordered by the `value_indices` in the `CSR`, resulting the same value order as the original CSC.

However, I still think removing `value_indices` in the COO will make our implementation not flexible. For example, if we want to sort the COO by column, we need to create a new sparse matrix.

Reply:
I am convinced to remove `value_indices` in the COO. Otherwise, it will make `mat.coo()` become more complex.